File: ProjectAshfall/ProjectAshfall/Models/Resources/Resource.swift
// Resource.swift
// ProjectAshfall
//
// Economy layer: resource types, costs, and a central inventory.

import Foundation

// MARK: - Resource Type

/// Every material that can be gathered, spent, or traded.
enum ResourceType: String, Codable, CaseIterable {
    case food
    case water
    case fuel
    case scrap
    case electronics
    case medicine
    case bioSamples
    case militaryComponents
    case powerCells
    case researchData

    /// Human-readable label for UI.
    var displayName: String {
        switch self {
        case .food:                return "Food"
        case .water:               return "Water"
        case .fuel:                return "Fuel"
        case .scrap:               return "Scrap"
        case .electronics:         return "Electronics"
        case .medicine:            return "Medicine"
        case .bioSamples:          return "Bio-Samples"
        case .militaryComponents:  return "Military Components"
        case .powerCells:          return "Power Cells"
        case .researchData:        return "Research Data"
        }
    }

    /// Asset name for the resource icon.
    var iconName: String {
        "icon_resource_\(rawValue)"
    }

    /// Base rarity weight — higher means harder to find in the world.
    var scarcity: Int {
        switch self {
        case .food, .water:                          return 1
        case .scrap, .fuel:                          return 2
        case .electronics, .medicine:                return 3
        case .bioSamples, .militaryComponents:       return 4
        case .powerCells, .researchData:             return 5
        }
    }
}

// MARK: - Resource Cost

/// A bill of materials used for building, upgrading, and crafting.
struct ResourceCost: Codable, Hashable {
    let type: ResourceType
    let amount: Int

    init(_ type: ResourceType, _ amount: Int) {
        self.type = type
        self.amount = amount
    }
}

// MARK: - Resource Inventory

/// Central store for the player's accumulated resources. Thread-safety is
/// the caller's responsibility (e.g. wrap in an actor at the system level).
final class ResourceInventory: Codable {

    // MARK: Storage

    private var storage: [ResourceType: Int]
    private var capacities: [ResourceType: Int]

    // MARK: Init

    init(defaultCapacity: Int = 9999) {
        storage = Dictionary(uniqueKeysWithValues: ResourceType.allCases.map { ($0, 0) })
        capacities = Dictionary(uniqueKeysWithValues: ResourceType.allCases.map { ($0, defaultCapacity) })
    }

    // MARK: Queries

    /// Current amount of a single resource.
    func amount(of type: ResourceType) -> Int {
        storage[type, default: 0]
    }

    /// Maximum storable amount for a resource type.
    func capacity(of type: ResourceType) -> Int {
        capacities[type, default: 9999]
    }

    /// Whether the inventory contains enough to cover every entry in `costs`.
    func canAfford(_ costs: [ResourceCost]) -> Bool {
        for cost in costs {
            if amount(of: cost.type) < cost.amount { return false }
        }
        return true
    }

    /// Returns a snapshot dictionary of all resource amounts.
    func allAmounts() -> [ResourceType: Int] {
        storage
    }

    // MARK: Mutations

    /// Adds `quantity` units, clamped to the current capacity.
    /// Negative values are silently ignored.
    @discardableResult
    func add(_ type: ResourceType, quantity: Int) -> Int {
        guard quantity > 0 else { return amount(of: type) }
        let cap = capacity(of: type)
        let current = storage[type, default: 0]
        storage[type] = min(cap, current + quantity)
        return storage[type, default: 0]
    }

    /// Spends `quantity` units if available.
    /// - Returns: `true` when the spend succeeded.
    @discardableResult
    func spend(_ type: ResourceType, quantity: Int) -> Bool {
        guard quantity > 0, amount(of: type) >= quantity else { return false }
        storage[type, default: 0] -= quantity
        return true
    }

    /// Attempts to deduct an array of costs atomically.
    /// Either all costs are deducted or none are.
    @discardableResult
    func spend(_ costs: [ResourceCost]) -> Bool {
        guard canAfford(costs) else { return false }
        for cost in costs {
            storage[cost.type, default: 0] -= cost.amount
        }
        return true
    }

    /// Sets the quantity of a resource directly (useful for save/load).
    func set(_ type: ResourceType, quantity: Int) {
        storage[type] = max(0, min(quantity, capacity(of: type)))
    }

    /// Increases the storage capacity for a given resource.
    func setCapacity(_ type: ResourceType, capacity: Int) {
        capacities[type] = max(0, capacity)
    }
}

File: src/models/Resource.ts
/**
 * Resource.ts — Resource types and inventory management for Project Ashfall.
 *
 * Tracks all resource currencies used for building, crafting, upgrading,
 * and mission rewards. Supports serialization for save/load.
 */

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

export enum ResourceType {
  Food = 'Food',
  Water = 'Water',
  Fuel = 'Fuel',
  Scrap = 'Scrap',
  Electronics = 'Electronics',
  Medicine = 'Medicine',
  BioSamples = 'BioSamples',
  MilitaryComponents = 'MilitaryComponents',
  PowerCells = 'PowerCells',
  ResearchData = 'ResearchData',
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

const RARE_RESOURCES: ReadonlySet<ResourceType> = new Set([
  ResourceType.BioSamples,
  ResourceType.MilitaryComponents,
  ResourceType.PowerCells,
  ResourceType.ResearchData,
]);

/** Returns `true` when the given resource is classified as rare. */
export function isRare(type: ResourceType): boolean {
  return RARE_RESOURCES.has(type);
}

/** Default amounts granted at game start. */
export const STARTING_RESOURCES: ReadonlyMap<ResourceType, number> = new Map([
  [ResourceType.Food, 500],
  [ResourceType.Water, 500],
  [ResourceType.Fuel, 200],
  [ResourceType.Scrap, 300],
  [ResourceType.Electronics, 50],
  [ResourceType.Medicine, 100],
  [ResourceType.BioSamples, 0],
  [ResourceType.MilitaryComponents, 0],
  [ResourceType.PowerCells, 0],
  [ResourceType.ResearchData, 0],
]);

// ---------------------------------------------------------------------------
// Serialization shape
// ---------------------------------------------------------------------------

export interface ResourceInventoryJSON {
  amounts: Record<string, number>;
}

// ---------------------------------------------------------------------------
// ResourceInventory
// ---------------------------------------------------------------------------

export class ResourceInventory {
  public amounts: Map<ResourceType, number>;

  constructor(initial?: Map<ResourceType, number>) {
    this.amounts = new Map<ResourceType, number>();

    // Seed every resource type with 0, then overlay provided values.
    for (const rt of Object.values(ResourceType)) {
      this.amounts.set(rt, 0);
    }

    if (initial) {
      for (const [rt, amt] of initial) {
        this.amounts.set(rt, amt);
      }
    }
  }

  // -----------------------------------------------------------------------
  // Factory helpers
  // -----------------------------------------------------------------------

  /** Create an inventory pre-loaded with the default starting amounts. */
  static createStartingInventory(): ResourceInventory {
    return new ResourceInventory(new Map(STARTING_RESOURCES));
  }

  // -----------------------------------------------------------------------
  // Accessors / mutators
  // -----------------------------------------------------------------------

  /** Get the current amount of a specific resource. */
  getAmount(type: ResourceType): number {
    return this.amounts.get(type) ?? 0;
  }

  /**
   * Add a positive `amount` of the given resource.
   * Negative values are silently ignored.
   */
  add(type: ResourceType, amount: number): void {
    if (amount <= 0) return;
    this.amounts.set(type, this.getAmount(type) + amount);
  }

  /**
   * Spend `amount` of the given resource.
   * @returns `true` if the spend succeeded, `false` if insufficient.
   */
  spend(type: ResourceType, amount: number): boolean {
    if (amount <= 0) return true;
    const current = this.getAmount(type);
    if (current < amount) return false;
    this.amounts.set(type, current - amount);
    return true;
  }

  /**
   * Check whether the inventory can cover **all** entries in `costs`.
   */
  canAfford(costs: Map<ResourceType, number>): boolean {
    for (const [rt, needed] of costs) {
      if (this.getAmount(rt) < needed) return false;
    }
    return true;
  }

  /**
   * Atomically deduct multiple resource costs.
   * @returns `true` if all costs were deducted, `false` (no-op) otherwise.
   */
  spendMultiple(costs: Map<ResourceType, number>): boolean {
    if (!this.canAfford(costs)) return false;
    for (const [rt, needed] of costs) {
      this.spend(rt, needed);
    }
    return true;
  }

  // -----------------------------------------------------------------------
  // Serialization
  // -----------------------------------------------------------------------

  toJSON(): ResourceInventoryJSON {
    const amounts: Record<string, number> = {};
    for (const [rt, amt] of this.amounts) {
      amounts[rt] = amt;
    }
    return { amounts };
  }

  static fromJSON(data: ResourceInventoryJSON): ResourceInventory {
    const map = new Map<ResourceType, number>();
    for (const [key, val] of Object.entries(data.amounts)) {
      if (Object.values(ResourceType).includes(key as ResourceType)) {
        map.set(key as ResourceType, val);
      }
    }
    return new ResourceInventory(map);
  }
}
